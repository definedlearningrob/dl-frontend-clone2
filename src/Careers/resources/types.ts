import { ApplicationStatus, OpportunityTypes } from '@graphql/dc/shared/types';
import { VirtualInternshipStatuses } from '@graphql/dc/students/types';
import { Partner } from '@graphql/dc/users/types';

import { CareerReviewSurveyQuestionOption } from '@dc/graphql/user/queries/lesson';

import {
  INSTITUTION_TYPES,
  OPPORTUNITY_APPLICATION_STATUS,
  SYNC_STATUS,
  VISIBILITY_SCOPE,
  APPLICATIONS_TYPE,
} from './enums';

type TCluster = {
  description: string;
  id: string;
  imageUrl: string;
  name: string;
  pathways: {
    id: string;
  };
  thumbnailUrl: string;
};

export type TOpportunityApplication = {
  id: string;
  appliedAt: string;
  updatedAt: string;
  lastChangedBy?: {
    uuid: string;
    name: string;
  };
  status: OPPORTUNITY_APPLICATION_STATUS;
  answers: {
    id: string;
    opportunityQuestionId: string;
    answer: string;
  }[];
  student: {
    uuid: string;
    fullName: string;
    schoolClasses: {
      uuid: string;
      name: string;
    }[];
  };
};

export type TOpportunityApplications = {
  nodes: TOpportunityApplication[];
  pagesCount: number;
  nodesCount: number;
};

export type TPathway = {
  cluster: TCluster;
  courses: {
    id: string;
  };
  description: string;
  id: string;
  imageUrl: string;
  name: string;
  thumbnailUrl: string;
};
export type TEntities = {
  uuid: string;
  name: string;
};

export type TLesson = {
  archivedAt: string;
  id: string;
  imageUrl: string;
  name: string;
  thumbnailUrl: string;
  type: string;
  progress: {
    submitted: number;
    total: number;
  };
  careerReviewSurvey: {
    performed: boolean;
  };
  description: {
    audience: string;
    goal: string;
    introduction: string;
    role: string;
    situation: string;
  };
};

export type TOpportunity = {
  applicationStatus: ApplicationStatus | null;
  archivedAt: string;
  createdAt: string;
  automaticAcceptance: boolean;
  availableSpots: number;
  creditsOutcomes: string;
  deadline: string | null;
  description: string;
  id: string;
  imageUrl: string;
  thumbnailUrl: string;
  isFavorite: boolean;
  isRecommended: boolean;
  location: string | null;
  name: string;
  hasPendingApplications: boolean;
  questions: {
    id: string;
    question: string;
    answer: string;
  }[];
  opportunityType: OpportunityTypes;
  opportunityApplication: { id: string; answers: { answer: string }[] };
  pathways: TPathway[];
  partner: Pick<Partner, 'id' | 'name'> | null;
  entities: TEntities[];
  periodEnd: string | null;
  periodStart: string | null;
  salaryInformation: string | null;
  tags: string[];
  visibilityScope: VISIBILITY_SCOPE;
  applications: TOpportunityApplications;
  virtualInternship: TVirtualInternship | null;
  imageFitToContainer: boolean;
};

export type TVirtualInternship = {
  id: string;
  roadmapItemsCount: number;
  status: VirtualInternshipStatuses;
  requiredExperiences: number;
  calendarLessons: TCalendarLesson[];
  experienceOpportunityLessons: ExperienceOpportunityLesson[];
  studentExperienceOpportunityLessons: TVirtualInternshipLesson[];
  readinessSkillsLessons: TVirtualInternshipLesson[];
  postExperienceLessons: TVirtualInternshipLesson[];
};

type ExperienceOpportunityLesson = Pick<
  TLesson,
  'type' | 'id' | 'name' | 'thumbnailUrl' | 'progress' | 'description'
>;

export type TVirtualInternshipLesson = Pick<
  TLesson,
  'name' | 'thumbnailUrl' | 'id' | 'progress' | 'careerReviewSurvey' | 'type'
>;

export type TCalendarLesson = Omit<TVirtualInternshipLesson, 'careerReviewSurvey'>;

export type TInstitution = {
  actMax: number | null;
  actMin: number | null;
  admissionRate: number | null;
  contact: TInstitutionContact;
  commonAppEnabled: boolean;
  hasApplied: boolean;
  type: INSTITUTION_TYPES;
  applicationType: APPLICATIONS_TYPE | null;
  applicationId: string | null;
  size?: string | null;
  sizeType: TInstitutionSizeType | null;
  sizeDescription: TInstitutionSizeDescription | null;
  cost: number | null;
  dates: TInstitutionDate[];
  thumbnailUrl: string | null;
  id: string;
  maxTeacherEval: number | null;
  minTeacherEval: number | null;
  isFavorite?: boolean;
  imageUrl: string | null;
  name: string;
  address: TInstitutionAddress;
  degrees: string[];
  description: string | null;
  satMathMax: number | null;
  satMathMin: number | null;
  satReadingMax: number | null;
  satReadingMin: number | null;
  studentFacultyRatio: number | null;
  commonAppApplicationUrl: string | null;
  isIpeds: boolean;
};

export const institutionSizeDescriptionMap = {
  LESS_THAN_500: '0-500',
  FROM_500_TO_1999: '500-1999',
  LESS_THAN_1000: '0-1000',
  FROM_1000_TO_2999: '1000-2999',
  FROM_2000_TO_4999: '2000-4999',
  FROM_3000_TO_9999: '3000-9999',
  FROM_5000_TO_9999: '5000-9999',
  MORE_THAN_10000: '10 000+',
};

export type TInstitutionSizeDescription = keyof typeof institutionSizeDescriptionMap;

export const institutionSizeTypeMap = {
  VERY_SMALL: 'Very Small',
  SMALL: 'Small',
  MEDIUM: 'Medium',
  LARGE: 'Large',
  VERY_LARGE: 'Very Large',
};

export type TInstitutionSizeType = keyof typeof institutionSizeTypeMap;

export type TInstitutionDate = {
  deadlineDate: string;
  decisionType: string;
  term: string;
};

type TInstitutionAddress = {
  city: string;
  state: string;
  stateCode: string;
  street: string;
  zip: string;
  area: {
    kind: string;
    type: string;
  };
};

type TInstitutionContact = {
  phone: string | null;
  urlAdmissions: string | null;
  urlApplications: string | null;
  urlFinancialAid: string | null;
  urlGeneral: string | null;
  urlNetPriceCalculator: string | null;
};

export type TSyncStatus = {
  lastSyncedAt: string | null;
  status: SYNC_STATUS;
};

export type TCollection = {
  id: string;
  name: string;
};

export type TCareerReviewSurveyQuestion = {
  answer: string[];
  id: string;
  options: CareerReviewSurveyQuestionOption[];
  question: string;
  type: string;
};
