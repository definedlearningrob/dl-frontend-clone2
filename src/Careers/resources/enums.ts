export enum CourseType {
  HIGH_SCHOOL = 'HIGH_SCHOOL',
  MIDDLE_SCHOOL = 'MIDDLE_SCHOOL',
}
export enum TagTypes {
  SYSTEM = 'SYSTEM',
  ENTITY = 'ENTITY',
}

export enum ResourceClass {
  ASSIGNMENT_SUBMISSION_FILE = 'ASSIGNMENT_SUBMISSION_FILE',
  ATTACHMENT = 'ATTACHMENT',
  CATALOG = 'CATALOG',
  COURSE = 'COURSE',
  EXTENSION_FIELD = 'EXTENSION_FIELD',
  EXTENSION_FIELD_FILE = 'EXTENSION_FIELD_FILE',
  LESSON = 'LESSON',
  PORTFOLIO_PROJECT = 'PORTFOLIO_PROJECT',
  PORTFOLIO_PROJECT_FILE = 'PORTFOLIO_PROJECT_FILE',
  PUBLIC_RESOURCE = 'PUBLIC_RESOURCE',
  RESUME = 'RESUME',
  SLIDE_BACKGROUND_IMAGE = 'SLIDE_BACKGROUND_IMAGE',
  SLIDE_IMAGE = 'SLIDE_IMAGE',
  SLIDE_VIDEO = 'SLIDE_VIDEO',
  TASK = 'TASK',
  TASK_FILE = 'TASK_FILE',
  TRACK = 'TRACK',
  UNIT = 'UNIT',
  VIDEO = 'VIDEO',
  BADGE = 'BADGE',
  ENTITY = 'ENTITY',
}

export enum AssetType {
  FILE = 'FILE',
  IMAGE = 'IMAGE',
  VIDEO = 'VIDEO',
  DC_LOGO = 'DC_LOGO',
  DC_ICON = 'DC_ICON',
  DL_LOGO = 'DL_LOGO',
  DL_ICON = 'DL_ICON',
}

export enum AssessmentType {
  HIGH_SCHOOL = 'HIGH_SCHOOL',
  MIDDLE_SCHOOL = 'MIDDLE_SCHOOL',
}

export enum AssessmentStatuses {
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
  FINISHED = 'FINISHED',
  IN_PROGRESS = 'IN_PROGRESS',
  NOT_STARTED = 'NOT_STARTED',
  PROCESSING_RESULTS = 'PROCESSING_RESULTS',
}

export enum ConversationContextTypes {
  ASSIGNMENT = 'ASSIGNMENT',
  LESSON = 'LESSON',
  COURSE = 'COURSE',
  CHECK_IN_ANSWER = 'CHECK_IN_ANSWER',
  ANNOUNCEMENT = 'ANNOUNCEMENT',
  OPPORTUNITY = 'OPPORTUNITY',
}

export enum Roles {
  ENTITY_ADMIN = 'ENTITY_ADMIN',
  SALES_ADMIN = 'SALES_ADMIN',
  SYSTEM_ADMIN = 'SYSTEM_ADMIN',
  TEACHER = 'TEACHER',
}

export enum EducationalSettingTypes {
  HIGH_SCHOOL = 'HIGH_SCHOOL',
  MIDDLE_SCHOOL = 'MIDDLE_SCHOOL',
}

export enum ReportLevels {
  USER = 'USER',
  SCHOOL_CLASS = 'SCHOOL_CLASS',
  ENTITY = 'ENTITY',
}

export enum UploadReportStatuses {
  NOT_STARTED = 'NOT_STARTED',
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
}

export enum ArchivableStatusTypes {
  ACTIVE = 'ACTIVE',
  ALL = 'ALL',
  ARCHIVED = 'ARCHIVED',
}

export enum ContentStatusesTypes {
  DRAFT = 'DRAFT',
  PUBLISHED = 'PUBLISHED',
}

export enum ServiceTypes {
  ALL = 'ALL',
  CAREERS = 'CAREERS',
  LEARNING = 'LEARNING',
}

export enum AssignmentSubmissionStatus {
  DRAFT = 'DRAFT',
  SUBMITTED = 'SUBMITTED',
}

export enum OPPORTUNITY_APPLICATION_STATUS {
  ACCEPTED = 'ACCEPTED',
  FINISHED = 'FINISHED',
  PENDING = 'PENDING',
  REJECTED = 'REJECTED',
  STARTED = 'STARTED',
  EXPIRED = 'EXPIRED',
}

export enum VIRTUAL_INTERNSHIP_STATUS {
  NOT_STARTED = 'NOT_STARTED',
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
}

export enum OPPORTUNITY_TYPE {
  APPRENTICESHIP = 'APPRENTICESHIP',
  CLINICAL_EXPERIENCE = 'CLINICAL_EXPERIENCE',
  INTERNSHIP = 'INTERNSHIP',
  JOB_SHADOW = 'JOB_SHADOW',
  OTHER = 'OTHER',
  PRACTICUM = 'PRACTICUM',
  PRE_APPRENTICESHIP = 'PRE_APPRENTICESHIP',
  VIRTUAL_INTERNSHIP = 'VIRTUAL_INTERNSHIP',
  WORKPLACE_TOUR = 'WORKPLACE_TOUR',
  JOB = 'JOB',
  COLLEGE_VISIT = 'COLLEGE_VISIT',
  SPEAKER = 'SPEAKER',
}

export enum VISIBILITY_SCOPE {
  ALL = 'ALL',
  ENTITY = 'ENTITY',
}

export enum INSTITUTION_TYPES {
  PRV_NFP_LT2 = 'Private not-for-profit, less-than 2-year',
  PRV_NFP_2 = 'Private not-for-profit, 2-year',
  PRV_NFP_4 = 'Private not-for-profit, 4-year or above',
  PRV_FP_LT2 = 'Private for-profit, less-than 2-year',
  PRV_FP_2 = 'Private for-profit, 2-year',
  PRV_FP_4 = 'Private for-profit, 4-year or above',
  PUB_LT2 = 'Public, less-than 2-year',
  PUB_2 = 'Public, 2-year',
  PUB_4 = 'Public, 4-year or above',
}

export enum US_STATES {
  ALASKA = 'Alaska',
  ALABAMA = 'Alabama',
  ARKANSAS = 'Arkansas',
  AMERICAN_SAMOA = 'American Samoa',
  ARIZONA = 'Arizona',
  CALIFORNIA = 'California',
  COLORADO = 'Colorado',
  CONNECTICUT = 'Connecticut',
  DISTRICT_OF_COLUMBIA = 'District Of Columbia',
  DELAWARE = 'Delaware',
  FLORIDA = 'Florida',
  GEORGIA = 'Georgia',
  GUAM = 'Guam',
  HAWAII = 'Hawaii',
  IOWA = 'Iowa',
  IDAHO = 'Idaho',
  ILLINOIS = 'Illinois',
  INDIANA = 'Indiana',
  KANSAS = 'Kansas',
  KENTUCKY = 'Kentucky',
  LOUISIANA = 'Louisiana',
  MASSACHUSETTS = 'Massachusetts',
  MARYLAND = 'Maryland',
  MAINE = 'Maine',
  MICHIGAN = 'Michigan',
  MINNESOTA = 'Minnesota',
  MISSOURI = 'Missouri',
  MISSISSIPPI = 'Mississippi',
  MONTANA = 'Montana',
  NORTH_CAROLINA = 'North Carolina',
  NORTH_DAKOTA = 'North Dakota',
  NEBRASKA = 'Nebraska',
  NEW_HAMPSHIRE = 'New Hampshire',
  NEW_JERSEY = 'New Jersey',
  NEW_MEXICO = 'New Mexico',
  NEVADA = 'Nevada',
  NEW_YORK = 'New York',
  OHIO = 'Ohio',
  OKLAHOMA = 'Oklahoma',
  OREGON = 'Oregon',
  PENNSYLVANIA = 'Pennsylvania',
  PUERTO_RICO = 'Puerto Rico',
  RHODE_ISLAND = 'Rhode Island',
  SOUTH_CAROLINA = 'South Carolina',
  SOUTH_DAKOTA = 'South Dakota',
  TENNESSEE = 'Tennessee',
  TEXAS = 'Texas',
  UTAH = 'Utah',
  VIRGINIA = 'Virginia',
  VIRGIN_ISLANDS = 'Virgin Islands',
  VERMONT = 'Vermont',
  WASHINGTON = 'Washington',
  WISCONSIN = 'Wisconsin',
  WEST_VIRGINIA = 'West Virginia',
  WYOMING = 'Wyoming',
}

export enum APPLICATIONS_TYPE {
  COMMON_APP = 'COMMON_APP',
  DIRECT = 'DIRECT',
}

export enum RECOMMENDER_TYPE {
  TEACHER = 'TEACHER',
  COUNSELOR = 'COUNSELOR',
}

export enum INSTITUTION_APPLICATION_STATUS {
  NOT_STARTED = 'NOT_STARTED',
  IN_PROGRESS = 'IN_PROGRESS',
  SUBMITTED = 'SUBMITTED',
  COMPLETED = 'COMPLETED',
  DOWNLOADED = 'DOWNLOADED',
}
export enum COMMON_APP_FORM_STATUS {
  NOT_STARTED = 'NOT_STARTED',
  IN_PROGRESS = 'IN_PROGRESS',
  SUBMITTED = 'SUBMITTED',
  COMPLETED = 'COMPLETED',
  DOWNLOADED = 'DOWNLOADED',
}

export enum APPLICATION_FORM_TYPE {
  COUNSELOR_EARLY_DECISION_2 = 'Counselor early decision 2',
  COUNSELOR_EARLY_DECISION = 'Counselor early decision',
  COUNSELOR_FEE_WAIVER = 'Counselor fee waiver',
  COUNSELOR_FINAL_REPORT = 'Counselor final report',
  COUNSELOR_MIDYEAR_REPORT = 'Counselor midyear report',
  COUNSELOR_OPTIONAL_REPORT_2 = 'Counselor optional report 2',
  COUNSELOR_OPTIONAL_REPORT = 'Counselor optional report',
  COUNSELOR_PROFILE = 'Counselor profile',
  COUNSELOR_RECOMMENDATION = 'Counselor recommendation',
  COUNSELOR_SCHOOL_REPORT = 'Counselor school report',
  COUNSELOR_SECONDARY_REPORT = 'Counselor secondary report',
  TEACHER_PROFILE = 'Teacher profile',
  TEACHER_RECOMMENDATION = 'Teacher recommendation',
  TEACHER_EVALUATION = 'Teacher evaluation',
}

export enum SIZE_TYPES {
  VERY_SMALL = 'VERY_SMALL', // Size for very small institutions
  SMALL = 'SMALL', // Size for small institutions
  MEDIUM = 'MEDIUM', // Size for medium institutions
  LARGE = 'LARGE', // Size for large institutions
  VERY_LARGE = 'VERY_LARGE', // Size for very large institutions
}

export enum COST_RANGES {
  LESS_THAN_5000 = 'LESS_THAN_5000', // Institution cost value between 0 and 5000
  FROM_5001_TO_10000 = 'FROM_5001_TO_10000', //Institution cost value between 5001 and 10000
  FROM_10001_TO_15000 = 'FROM_10001_TO_15000', //Institution cost value between 10001 and 15000
  FROM_15001_TO_20000 = 'FROM_15001_TO_20000', //Institution cost value between 15001 and 20000
  MORE_THAN_20000 = 'MORE_THAN_20000', //Institution cost value higher than 20000
}

export enum LOCALE_STORAGE_KEY {
  APPLICATION_REFRESH_TIME = 'APPLICATION_REFRESH_TIME',
}

export enum CAREER_REVIEW_SURVEY_ANSWER_CONTEXT_TYPES {
  COURSE = 'COURSE',
  VIRTUAL_INTERNSHIP = 'VIRTUAL_INTERNSHIP',
  ASSESSMENT = 'ASSESSMENT',
}

export enum SYNC_STATUS {
  COMPLETED = 'COMPLETED',
  IN_PROGRESS = 'IN_PROGRESS',
  FAILED = 'FAILED',
}

export enum ASSIGNMENT_SUBMISSION_TYPES {
  COURSE = 'COURSE',
  VIRTUAL_INTERNSHIP = 'VIRTUAL_INTERNSHIP',
}

export enum ASSESSMENT_TYPES {
  HIGH_SCHOOL = 'HIGH_SCHOOL',
  MIDDLE_SCHOOL = 'MIDDLE_SCHOOL',
}

export enum ASSESSMENT_ATTEMPT_STATUS {
  IN_PROGRESS = 'IN_PROGRESS',
  FINISHED = 'FINISHED',
  REJECTED = 'REJECTED',
  PROCESSING_RESULTS = 'PROCESSING_RESULTS',
  FAILED = 'FAILED',
}

export enum ASSESSMENT_STATUSES {
  COMPLETED = 'COMPLETED',
  NOT_STARTED = 'NOT_STARTED',
  IN_PROGRESS = 'IN_PROGRESS',
}
