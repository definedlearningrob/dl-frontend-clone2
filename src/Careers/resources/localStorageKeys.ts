export const LOCAL_STORAGE_COURSE_ID = 'course-id';
