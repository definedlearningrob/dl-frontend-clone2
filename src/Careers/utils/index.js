const isObject = (object) => object !== null && typeof object === 'object';

export const isEqual = (object1, object2) => {
  if (object1 === object2) return true;

  if (
    object1 === null ||
    object2 === null ||
    typeof object1 !== 'object' ||
    typeof object2 !== 'object'
  )
    return false;

  const keys1 = Object.keys(object1);
  const keys2 = Object.keys(object2);

  if (keys1.length !== keys2.length) {
    return false;
  }

  for (let key of keys1) {
    const val1 = object1[key];
    const val2 = object2[key];
    const areObjects = isObject(val1) && isObject(val2);
    if (
      // eslint-disable-next-line
      areObjects && !isEqual(val1, val2) ||
      // eslint-disable-next-line
      !areObjects && val1 !== val2
    ) {
      return false;
    }
  }

  return true;
};

export const getByStringKey = (object, key) => {
  const keys = key.split('.');

  return keys.reduce((acc = {}, key) => acc[key], object);
};
