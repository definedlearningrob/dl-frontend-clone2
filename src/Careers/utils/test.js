/* eslint-disable react/prop-types */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { createMemoryHistory } from 'history';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { Router, Route } from 'react-router-dom';

import i18n from '@dc/i18n';

export function renderWithRouter(
  ui,
  { route = '/', routePath = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
  const Wrapper = ({ children }) => (
    <Router history={history}>
      <I18nextProvider i18n={i18n}>
        <Route path={routePath} render={() => children} />
      </I18nextProvider>
    </Router>
  );

  return {
    ...render(ui, { wrapper: Wrapper }),
    history,
  };
}

export function renderWithReduxProvider(
  ui,
  {
    initialState = {
      session: {
        loginError: {},
        user: null,
      },
    },
    store = configureMockStore([thunk])(initialState),
  } = {}
) {
  const Wrapper = ({ children }) => (
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </Provider>
  );

  return {
    ...render(ui, { wrapper: Wrapper }),
  };
}

export function renderWithRouterAndReduxProvider(
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
    initialState = {
      session: {
        loginError: {},
        user: {},
      },
    },
    store = configureMockStore([thunk])(initialState),
  } = {},
  ...options
) {
  const Wrapper = ({ children }) => (
    <Provider store={store}>
      <Router history={history}>
        <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
      </Router>
    </Provider>
  );

  return {
    ...render(ui, { wrapper: Wrapper, ...options }),
    history,
  };
}

export function renderWithI18N(ui) {
  const Wrapper = ({ children }) => <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;

  return {
    ...render(ui, { wrapper: Wrapper }),
  };
}
