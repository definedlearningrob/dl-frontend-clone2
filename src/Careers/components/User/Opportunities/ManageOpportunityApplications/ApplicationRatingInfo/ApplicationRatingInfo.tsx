import { useTranslation, Trans } from 'react-i18next';
import cx from 'classnames';

import { TOpportunityApplication } from '@dc/resources/types';
import { OPPORTUNITY_APPLICATION_STATUS } from '@dc/resources/enums';
import { ReactComponent as ThumbsUpIcon } from '@dc/assets/icons/thumbs-up.svg';
import { ReactComponent as ThumbsDownIcon } from '@dc/assets/icons/thumbs-down.svg';
import { AcceptedStatuses } from '@dc/graphql/user/hooks/useUpdateOpportunityApplication';

import { ReactComponent as ClearIcon } from '@shared/assets/icons/clear.svg';
import Dropdown from '@shared/components/Dropdown/Dropdown';
import Icon from '@shared/components/Icon/Icon';

type Props = {
  application: TOpportunityApplication;
  onStatusChange: (status: AcceptedStatuses) => void;
  onResetApplication: () => void;
};

export const ApplicationRatingInfo = ({
  application,
  onStatusChange,
  onResetApplication,
}: Props) => {
  const applicationStatus = application.status;
  const { t } = useTranslation();
  const isApplicationAccepted = applicationStatus === OPPORTUNITY_APPLICATION_STATUS.ACCEPTED;
  const isApplicationRejected = applicationStatus === OPPORTUNITY_APPLICATION_STATUS.REJECTED;

  const statusIconClassName = cx('p-xs rounded-full', {
    'bg-success-100 text-success-500': isApplicationAccepted,
    'bg-danger-100 text-danger-500': isApplicationRejected,
  });
  const handleDropdownOption = () =>
    onStatusChange(
      isApplicationAccepted
        ? OPPORTUNITY_APPLICATION_STATUS.REJECTED
        : OPPORTUNITY_APPLICATION_STATUS.ACCEPTED
    );

  return (
    <>
      <Icon
        className={statusIconClassName}
        icon={isApplicationAccepted ? <ThumbsUpIcon /> : <ThumbsDownIcon />}
      />
      <div>
        <h5 className='text-primary-500 font-medium leading-lg text-xxs xxxl:text-sm mb-0'>
          {isApplicationAccepted
            ? t('opportunityManageApplications.details.acceptedInfo')
            : t('opportunityManageApplications.details.rejectedInfo')}
        </h5>
        <span className='text-neutral-700 text-xxs leading-lg mb-0 xxxl:text-xs'>
          <Trans
            i18nKey={
              isApplicationAccepted
                ? 'opportunityManageApplications.details.statusAccepted'
                : 'opportunityManageApplications.details.statusRejected'
            }
            values={{
              studentName: application.student.fullName,
              teacherName: application.lastChangedBy?.name,
            }}
          />
        </span>
      </div>
      <div className='ms-auto hover:rounded-xs hover:border hover:border-primary-500 focus:outline-1 focus:outline-primary-500 focus:outline-offset-1'>
        <Dropdown position='top'>
          <Dropdown.Dropdown>
            <Dropdown.Trigger />
            <Dropdown.Options>
              <Dropdown.Option onClick={handleDropdownOption}>
                <Icon
                  className='text-neutral-800'
                  icon={isApplicationAccepted ? <ThumbsDownIcon /> : <ThumbsUpIcon />}
                  size='xs'
                />
                <span className='text-neutral-800'>
                  {isApplicationAccepted
                    ? t('opportunityManageApplications.details.actionReject')
                    : t('opportunityManageApplications.details.actionAccept')}
                </span>
              </Dropdown.Option>
              <Dropdown.Option onClick={onResetApplication}>
                <Icon className='text-danger-600' icon={<ClearIcon />} size='xs' />
                <span className='text-danger-600'>
                  {t('opportunityManageApplications.details.actionCancel')}
                </span>
              </Dropdown.Option>
            </Dropdown.Options>
          </Dropdown.Dropdown>
        </Dropdown>
      </div>
    </>
  );
};
