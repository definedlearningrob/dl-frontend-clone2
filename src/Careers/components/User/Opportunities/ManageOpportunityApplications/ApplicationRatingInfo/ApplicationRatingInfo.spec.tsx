import { screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';

import { renderWithRouterAndReduxProvider } from '@dc/utils/test';
import { ApplicationRatingInfo } from '@dc/components/User/Opportunities/ManageOpportunityApplications/ApplicationRatingInfo/ApplicationRatingInfo';
import { OPPORTUNITY_APPLICATION_STATUS } from '@dc/resources/enums';

const application = {
  id: '1',
  appliedAt: '2023-05-01',
  updatedAt: '2023-05-02',
  status: OPPORTUNITY_APPLICATION_STATUS.PENDING,
  answers: [{ id: '1', opportunityQuestionId: '1', answer: 'To sleep better!', step: 1 }],
  lastChangedBy: {
    uuid: '123',
    name: 'Teacher Name',
  },
  student: {
    uuid: '123',
    fullName: 'Applicant Student 1',
    schoolClasses: [{ uuid: 'class-uuid', name: 'Class name' }],
  },
};

describe('ApplicationRatingInfo', () => {
  const resetSpy = jest.fn();
  const statusChangeSpy = jest.fn();

  it('should render correctly for rejected application', () => {
    const { container } = renderWithRouterAndReduxProvider(
      <ApplicationRatingInfo
        application={application}
        onResetApplication={resetSpy}
        onStatusChange={statusChangeSpy}
      />
    );

    expect(screen.getByRole('heading', { level: 5, name: 'Application rejected' }));
    expect(screen.getByText("Applicant Student 1's application was rejected by Teacher Name."));

    expect(container).toMatchSnapshot();
  });

  it('should render correctly for accepted application', () => {
    renderWithRouterAndReduxProvider(
      <ApplicationRatingInfo
        application={{ ...application, status: OPPORTUNITY_APPLICATION_STATUS.ACCEPTED }}
        onResetApplication={resetSpy}
        onStatusChange={statusChangeSpy}
      />
    );

    expect(screen.getByRole('heading', { level: 5, name: 'Application accepted' }));
    expect(screen.getByText("Applicant Student 1's application was accepted by Teacher Name."));
  });

  // skipped as I was not able to make the dropdown to show up, revisit after new dropdown is used in this component
  it.skip('should allow changing the status by using context menu', async () => {
    renderWithRouterAndReduxProvider(
      <div id='dropdown-portal-container'>
        <ApplicationRatingInfo
          application={application}
          onResetApplication={resetSpy}
          onStatusChange={statusChangeSpy}
        />
      </div>
    );

    userEvent.click(screen.getByRole('button'));
    userEvent.click(await screen.findByText('Change to accepted'));
  });
});
