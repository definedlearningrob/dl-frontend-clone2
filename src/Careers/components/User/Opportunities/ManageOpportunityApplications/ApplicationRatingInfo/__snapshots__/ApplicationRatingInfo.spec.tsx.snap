// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationRatingInfo should render correctly for rejected application 1`] = `
<div>
  <div
    class="icon p-xs rounded-full"
    data-testid="icon"
  >
    <div />
  </div>
  <div>
    <h5
      class="text-primary-500 font-medium leading-lg text-xxs xxxl:text-sm mb-0"
    >
      Application rejected
    </h5>
    <span
      class="text-neutral-700 text-xxs leading-lg mb-0 xxxl:text-xs"
    >
      Applicant Student 1's application was rejected by Teacher Name.
    </span>
  </div>
  <div
    class="ms-auto hover:rounded-xs hover:border hover:border-primary-500 focus:outline-1 focus:outline-primary-500 focus:outline-offset-1"
  >
    <div
      class="dropdown-portal-container"
    >
      <div
        class="dropdown-portal-options__trigger"
        data-testid="dropdown-trigger-container"
      >
        <button
          class="dropdown-portal-options__trigger-icon-wrapper"
        >
          <div
            class="icon dropdown-portal-options__trigger-icon -sm-size"
            data-testid="icon"
          >
            <div />
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
